### PR TITLE
additional tests for redeem rewards

### PR DESCRIPTION
**Pull request by piotr-iohk**
# Issue Number

# Overview

- 34f17f2237b562634ad3015eb03d9f1390382243
  additional tests for redeem rewards
  
- ef15832b62b3cd341356d9a45ca01dd768eaf658
  Add more reward wallets
  
- 85a07e7042ae3c940366827907f879358de07391
  Additional cases to cover not enough fee and not enough money on withdrawals

# Comments

in progress...

### CODE REVIEW

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 1]
```diff
@@ -1396,6 +1410,67 @@ spec = do
             [ expectResponseCode HTTP.status403
             , expectErrorMessage errMsg403NotAShelleyWallet
             ]
+
+    it "SHELLEY_TX_REDEEM_06a - Can't redeem rewards if utxo = 0 from other" $ \ctx -> do
```
Comment (paweljakubas) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:1422]: :+1:

**Review by paweljakubas — APPROVED** [review]
LGTM!

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 2]
```diff
@@ -1255,47 +1257,60 @@ spec = do
...
+                        , ("redeem sending less than UTxO", 100 * oneThousandAda - oneAda)
+                        , ("redeem sending more than rewards", oneMillionAda + oneAda)
+                        , ("redeem sending less than rewards", oneMillionAda - oneAda)
+                        ]
+    describe "SHELLEY_TX_REDEEM_01 - Can redeem rewards from self" $ do
+
+      forM_ rewardsMatrix $ \(name, amt) -> it name $ \ctx -> do
         (wSrc,_) <- rewardWallet ctx
         addr:_ <- fmap (view #id) <$> listAddresses @n ctx wSrc
+        print wSrc
```
Comment (KtorZ) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]: Whoops :point_up:

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 3]
```diff
@@ -1255,47 +1257,60 @@ spec = do
             , expectErrorMessage (errMsg404NoWallet wid)
             ]
 
-    it "SHELLEY_TX_REDEEM_01 - Can redeem rewards from self" $ \ctx -> do
+    -- Reward wallets have 100k ADA UTxO and 1M ADA rewards,
+    -- fixtureWalletsWith have 100k ADA UTxO too
+    -- Verifying withdrawals for some boundary values
+    let rewardsMatrix = [ ("redeem sending exactly rewards", oneMillionAda)
+                        , ("redeem sending exactly UTxO", 100 * oneThousandAda)
+                        , ("redeem sending more than UTxO", 100 * oneThousandAda + oneAda)
```
Comment (KtorZ) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]: Isn't this case redundant with the first one :grimacing:  ?

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 4]
```diff
@@ -1255,47 +1257,60 @@ spec = do
...
+                        , ("redeem sending less than UTxO", 100 * oneThousandAda - oneAda)
+                        , ("redeem sending more than rewards", oneMillionAda + oneAda)
+                        , ("redeem sending less than rewards", oneMillionAda - oneAda)
```
Comment (KtorZ) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]: Isn't this case redundant with the 2nd and 4th ones :grimacing:  ?

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs (hunk 3 above)
Comment (piotr-iohk) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]: It depends on how you look at it :) 
 - First is sending amt = rewards when utxo is < rewards
 - This one is send amt = utxo + 1 when utxo is < rewards

They are different "edge cases" for me... (but you know implementation details better ofc)
So if you tell me we can safely drop one we can do it.

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs (hunk 4 above)
Comment (piotr-iohk) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]: I guess not?

Perhaps it's more like 3rd...

**Comment by paweljakubas** [comment]
after running only redeem test suits it happened to take `744 seconds` to finish. I trimmed edge cases as they could be in the next PR put on unit test level. Also changed titles in some test to better convey message and nature of tests. After trimming the testsuites halved in time execution

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]: addressed in https://github.com/input-output-hk/cardano-wallet/pull/1987/commits/ff01561b383b3e0989700fb54a2ff5ae0b12a4da

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs (hunk 4 above)
Comment (paweljakubas) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]: addressed in https://github.com/input-output-hk/cardano-wallet/pull/1987/commits/ff01561b383b3e0989700fb54a2ff5ae0b12a4da

**Review by KtorZ — APPROVED** [review]
